overlord/snapstate/component.go

```diff
@@ -0,0 +1,227 @@
+	if si.RealName == "" {
+		return nil, fmt.Errorf(
+			"internal error: snap name to install component %q not provided",
+			path)
+	}
+
+	if snapInstance == "" {
+		snapInstance = si.RealName
+	}
+
+	deviceCtx, err := DeviceCtxFromState(st, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var snapst SnapState
+	err = Get(st, snapInstance, &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	channel, err = resolveChannel(snapInstance, snapst.TrackingChannel, channel,
+		deviceCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read ComponentInfo
+	compInfo, _, err := backend.OpenComponentFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check snap name matches
+	if compInfo.Component.SnapName != si.RealName {
+		return nil, fmt.Errorf(
+			"component snap name %q does not match real snap name %q",
+			compInfo.Component.SnapName, si.RealName)
+	}
+
+	// Read snap.Info from the installed snap
+	info, err := readInfo(snapInstance, si, withAuxStoreInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	providerContentAttrs := defaultProviderContentAttrs(st, info, nil)
+	snapsup := &SnapSetup{
+		Base:               info.Base,
+		Prereq:             getKeys(providerContentAttrs),
+		PrereqContentAttrs: providerContentAttrs,
+		SideInfo:           si,
+		SnapPath:           path,
+		Channel:            channel,
+		Flags:              flags.ForSnapSetup(),
+		Type:               info.Type(),
+		Version:            info.Version,
+		PlugsOnly:          len(info.Slots) == 0,
+		InstanceKey:        info.InstanceKey,
+	}
+
+	compSetup := &ComponentSetup{
+		CompSideInfo:    csi,
+		CompPath:        path,
+		SnapInstanceKey: info.InstanceKey,
+		SnapType:        info.Type(),
+		SnapRevision:    info.Revision,
+	}
+	removeComponentPath := true
+	return doInstallComponent(st, &snapst, compSetup, snapsup, path, removeComponentPath, "")
+}
+
+// doInstallComponent assumes that the owner snap is already installed.
+func doInstallComponent(st *state.State, snapst *SnapState, compSetup *ComponentSetup,
+	snapsup *SnapSetup, path string, removeComponentPath bool,
+	fromChange string) (*state.TaskSet, error) {
+
+	// TODO check for experimental flag that will hide temporarily components
+
+	snapSi := snapsup.SideInfo
+	compSi := compSetup.CompSideInfo
+
+	if snapst.IsInstalled() && !snapst.Active {
+		return nil, fmt.Errorf("cannot install component %q for disabled snap %q",
+			compSi.Component, snapSi.RealName)
+	}
+
+	// Check parallel constraints, if applicable
+	if err := isParallelInstallable(snapsup); err != nil {
+		return nil, err
+	}
+
+	// TODO extend conflict checks to components, this will check only for
+	// snaps conflicts (installation of a component as a snap gets the same
+	// conflicts as if we were installing the snap - which might be
+	// overkill and needs to be revisited).
+	if err := checkChangeConflictIgnoringOneChange(st, snapsup.InstanceName(),
```

Comment:
this won't really work well unless we attach snap-setup to the tasks, which anyway I think makes sense as I said previously

Reply:
I've now attached SnapSetup to ComponentSetup (although not really sure how is this check affected as we take only the instance name?)